test/unit/caret_test.js

```diff
@@ -0,0 +1,129 @@
+import sinon from 'sinon';
+import {expect} from 'chai';
+import installCaret from '../../src/caret';
+
+describe('Caret', () => {
+  describe('with no document', () => {
+    it('throws an error', () => {
+      expect(installCaret.bind(this, null)).to.throw(/Caret does not have access to document/);
+    });
+  });
+
+  describe('with no native caret', () => {
+    it('throws an error', () => {
+      expect(installCaret.bind(this, {
+        createElement: () => { return {}; }
+      })).to.throw(/Caret unknown input selection capabilities/);
+    });
+  });
+
+  describe('with selectionStart', () => {
+    let getCaret, setCaret;
+
+    beforeEach(() => {
+      const caret = installCaret({
+        createElement: () => { return { selectionStart: true }; }
+      });
+      getCaret = caret.getCaret;
+      setCaret = caret.setCaret;
+    });
+
+    it('getCaret calls selectionStart and selectionEnd', () => {
+      const fakeElement = {
+        selectionStart: 4,
+        selectionEnd: 8
+      };
+      expect(getCaret(fakeElement)).to.eql({
+        start: 4,
+        end: 8
+      });
+    });
+
+    it('setCaret sets selectionStart and selectionEnd', () => {
+      const fakeElement = {
+        selectionStart: 4,
+        selectionEnd: 8
+      };
+      setCaret(fakeElement, 1, 5);
+
+      expect(fakeElement).to.eql({
+        selectionStart: 1,
+        selectionEnd: 5
+      });
+    });
+  });
+
+  describe('with document.selection', function() {
+    let getCaret, setCaret;
+
+    beforeEach(() => {
+      const caret = installCaret({
+        createElement: () => { return {}; },
+        selection: {
+          createRange: () => {
+            return {
+              duplicate: () => {
+                const selectedText = this.selectedText;
```

Comment:
The context of these `this.selectedText` gets confusing in this section. In this one it is referring to the `this` inside the callback for `describe`. Inside the `TextRange` class it's an instance of `TextRange`. Can we move `TextRange` to a support file or the top of this one?